plugins/filter_kubernetes/kube_meta.c

```diff
@@ -226,6 +344,12 @@ static int get_meta_info_from_request(struct flb_kube *ctx,
         return -1;
     }
 
+    ret = refresh_token_if_needed(ctx);
+    if (ret == -1) {
+        flb_plg_error(ctx->ins, "failed to refresh token");
+        return -1;
+    }
```

Comment:
Since EKS is running the fluentbit as systemd process, [flb_kube_meta_init](https://github.com/fluent/fluent-bit/blob/37aa680d32384c1179f02ee08a5bef4cd278513e/plugins/filter_kubernetes/kube_meta.c#L1237) doesn't invoke get_http_auth_header(..). This is because get_local_pod_info [returns](https://github.com/fluent/fluent-bit/blob/37aa680d32384c1179f02ee08a5bef4cd278513e/plugins/filter_kubernetes/kube_meta.c#L107) before invoking get_http_auth_header().

Therefore after init() when fluentbit tries to get pod info by invoking get_meta_info_from_request(), refresh_token_if_needed() wouldn't do anything since ctx->kube_token_create is not initialized yet.

My suggestion would be instead of checking if ctx->kube_token_create is zero and refreshing the token, I would suggest to check `if (ctx->kube_token_command) and (kube_token_create is zero or expired)` then generate the token. Therefore whoever needs to get the token, they can just invoke refresh_token_if_needed and need not worry about whether get_http_auth_header is invoked or not (to initialize the token).

Reply:
Hello,

So for this version of code, I've changed that part of logic: https://github.com/fluent/fluent-bit/pull/3677/files#diff-ac1b66f45abf086e63ffcc7273974409a88ba95a753ba04d2cc8c5d2d0486fe6L111-R258. If users enter a command, it will not return by FLB_FALSE here: https://github.com/fluent/fluent-bit/blob/37aa680d32384c1179f02ee08a5bef4cd278513e/plugins/filter_kubernetes/kube_meta.c#L107.

Reply:
I see. Thanks for your exploration